packages/record-data/addon/-private/ordered-set.ts

```diff
@@ -29,12 +125,13 @@ export default class EmberDataOrderedSet<T> extends EmberOrderedSet<T> {
     return this;
   }
 
-  deleteWithIndex(obj: T | null, idx?: number): boolean {
+  deleteWithIndex(obj: RelationshipRecordData | null, idx?: number): boolean {
     let guid = guidFor(obj);
+    assert(`Expected ${obj} to have an clientId`, typeof guid === 'string' && guid !== '');
     let presenceSet = this.presenceSet;
     let list = this.list;
 
-    if (presenceSet[guid] === true) {
+    if (presenceSet[guid] === obj) {
```

Comment:
This seems like a major change?  We are relying on referential equality now, which is different than if we have two instances with the same `guid`.